framework/connection.py

```diff
@@ -0,0 +1,87 @@
+"""
+Module to connect to remote host
+"""
+
+import logging
+
+from common_ci_utils.connection import Connection
+from framework import config
+from paramiko.auth_handler import AuthenticationException, SSHException
+
+log = logging.getLogger(__name__)
+
+
+class SSHConnection:
+    """
+    A class that connects to remote host
+    """
+
+    _instance = None
+
+    def __new__(cls, *args, **kwargs):
+        if not cls._instance:
+            cls._instance = super(SSHConnection, cls).__new__(cls)
+            cls._instance._conn = None
+        return cls._instance
+
+    def __init__(self):
+        """
+        Get connection to host
+
+        Raises:
+            authException: In-case of authentication failed
+            sshException: In-case of ssh connection failed
+
+        """
+        # Initialize the connection only if it hasn't been created yet
+        if not self._conn:
+            self.host = config.ENV_DATA["noobaa_sa_host"]
+            self.user = config.ENV_DATA["user"]
+            self.password = config.ENV_DATA.get("password")
+            self.private_key = config.ENV_DATA.get("private_key")
+            try:
+                if self.private_key:
+                    self._conn = Connection(
+                        host=self.host,
+                        user=self.user,
+                        private_key=self.private_key,
+                    )
+                else:
+                    self._conn = Connection(
+                        host=self.host,
+                        user=self.user,
+                        password=self.password,
+                    )
+            except AuthenticationException as authException:
+                log.error(f"Authentication failed: {authException}")
+                raise authException
+            except SSHException as sshException:
+                log.error(f"SSH connection failed: {sshException}")
+                raise sshException
+
+    def get_connection(self):
+        """
+        Get connection to host
+
+        Returns:
+            paramiko.client: Paramiko SSH client connection to host
+
+        """
+        return self._conn
+
+    def close_connection(self):
+        """
+        Closes SSH connection
+        """
+        if self._conn:
+            self._conn.close()
+            self._conn = None
+
+
+ssh_conn = SSHConnection()
```

Comment:
I still don't like the approach of setting up connection here on module level. As if you will load this module, python will need to always initiate the connection. That's why you have an connection class as singleton, you can access the same object only from places you need, especially from functions or class, but only when it's needed.

Singleton should init connection once and keep pointer to the same object all the time, even though you use it from different modul.

If you want to have sessionfinish hook to close the connection, you can maybe implement as part of the class for singleton classmethod:

```python
@classmethod
    def destroy(cls):
        if cls._instance and cls._instance._conn:
            cls._instance._conn.close()
            cls._instance = None
```

You can try to do pytest with --collect-only option and don't have currently set the data for connection in config if you won't have any noobaa SA instance running. In this case I think that your approach will fail on the initiating a ssh_con here - which we should not fail.

Reply:
done